### ISSUE TITLE
Fix for obscure RuntimeError: maximum recursion limit reached

### ISSUE THREAD

**Issue by apollo13** [ansible/ansible#10521]
Hi @jimi-c,

since my paste on IRC most likely already expired, this should be the fix for what we've been discussing today:

```
diff --git a/v2/ansible/playbook/base.py b/v2/ansible/playbook/base.py
index e32da5d..d5ab3d5 100644
--- a/v2/ansible/playbook/base.py
+++ b/v2/ansible/playbook/base.py
@@ -290,7 +290,7 @@ class Base:
         raise AttributeError("attribute not found in %s: %s" % (self.__class__.__name__, needle))

     def __setattr__(self, needle, value):
-        if hasattr(self, '_attributes') and needle in self._attributes:
+        if '_attributes' in self.__dict__ and needle in self._attributes:
             self._attributes[needle] = value
         else:
             super(Base, self).__setattr__(needle, value)
```

The issue was the assignment of stuff in `__init__` which goes through `__setattr__` and the `hasattr` check in there causes an infinite loop in `__getattr__`. It might not be the best fix, maybe `__getattr__` could also do with a bit more defensive coding, but it should work… (It does work for me at least, if it doesn't fix your issue, please uploading the whole play+roles and I'll throw gdb at it).

**Comment by jimi-c** [comment]
Thanks! We had narrowed it down to setattr too, but were thinking about making things proper python properties so as to be more pythonic. If that ends up going south, we'll use this instead.

**Comment by apollo13** [comment]
Perfect, I am all in for more pythonic code :) Out of curiosity, did you figure out why the error was only raised once, gdb clearly showed that the recursion limit was hit roughly somewhere <100 times before the output was generated once, so something was swallowing the error.

**Comment by abadger** [comment]
I didn't look specifically but I suspect that the garbage collector might be catching the recursion error and then rethrowing it as a warning instead of an error.  I think the default python settings make warnings thrown from the same code only display once instead of every time they are hit.

**Comment by jimi-c** [comment]
Just merged in 095990b4d8dcd93e65b188fb9ffeb37b1d3b09e5, so I'll close this as no longer being needed. @apollo13, if you see any further problems with this code, let us know. Thanks!

**Comment by apollo13** [comment]
Looks good so far (ie the error is gone). Personally I'd probably use a `metaclass/__new__` to collect the `Attribute`-classes and then install descriptors instead of that getter/setter magic. If you are not averse to such an approach I might give it a try this weekend.

**Comment by abadger** [comment]
I'm honestly not sure.  I'm not happy with the hackiness of some of my implementation via properties (I need to put in a comment to explain why the methods are marked `@staticmethod` and yet take self as argument 2) but python properties are generally very well understood and this just builds on that existing knowledge a tiny bit.

metaclasses are a whole 'nother area of python that few people have occasion to use so it raises the barrier to entry for reading and understanding the code quite a bit.  OTOH, even though the practice of using metaclasses is unfamiliar, the theory of metaclasses and `__new__` are pretty logical.  So an implementation using those might be acceptable.  I just wouldn't want to sacrifice too much in the way of familiarity even if it does get rid of an ugly hack.

Truthfully, using `__setattr__`` and`**getattr**``` seems like the right level to solve this except that, as we saw, it was prone to error conditions that were hard to debug.  The new property code at least alleviates that.

So bottom line -- if you're really wanting to play with a metaclass implementation you're welcome to but I'm not sure if we'd accept it.  Even a very clean implementation would involve tradeoffs between using something that people understand (properties) vs something that most python programmers never have to use (metaclasses).

**Comment by apollo13** [comment]
Ok, I'll give it a try. If you have to explain why `staticmethod` is needed, that is probably a sign of unpythonic code. Since those are helper functions anyways, just move them out of the class or into `_initialize_base_attributes`.

**Comment by abadger** [comment]
Yeah, I'd considered that yesterday.  The thing is that they really are
methods. It's just that python wants to bind then to the class instance
twice.  Making them staticmethods prevents one of those from adding self to
the signature.

That's part of why I call this a hack.  But like I say, the hack is
approachable by general python programmers in a way that metaclasses in
general are not so we may or may not take code implemented via metaclass.
Depends on how clear that code is to people who don't know about
metaclasses I think.
